From d1ee1e1e4907d2de913337a876d20bc5dbabfbc4 Mon Sep 17 00:00:00 2001
From: Heinrich Toews <heinrich.toews@wago.com>
Date: Thu, 18 Oct 2018 11:22:29 +0000
Subject: [PATCH 133/144] fs: tftp: add range checking for blocksize option

Tests showed that only blocksize values between 30 and 1468 octets
are working. So add a verification of this value here.

Signed-off-by: Heinrich Toews <heinrich.toews@wago.com>
---
 fs/tftp.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/fs/tftp.c b/fs/tftp.c
index def5aaa..b82d8ba 100644
--- a/fs/tftp.c
+++ b/fs/tftp.c
@@ -70,6 +70,9 @@
 #define STATE_DONE	8
 
 #define TFTP_BLOCK_SIZE		512	/* default TFTP block size */
+#define TFTP_MIN_BLOCK_SIZE	30	/* this range was empirically determined */
+#define TFTP_MAX_BLOCK_SIZE	1468
+
 #define TFTP_FIFO_SIZE		4096
 
 #define TFTP_ERR_RESEND	1
@@ -388,6 +391,15 @@ static struct file_priv *tftp_do_open(struct device_d *dev,
 
 	priv = xzalloc(sizeof(*priv));
 
+	if (global_tftp_block_size < TFTP_MIN_BLOCK_SIZE ||
+			global_tftp_block_size > TFTP_MAX_BLOCK_SIZE) {
+		pr_err("error: blocksize is unvalid (use: %d-%d)\n",
+				TFTP_MIN_BLOCK_SIZE,
+				TFTP_MAX_BLOCK_SIZE);
+		ret = -EINVAL;
+		goto out;
+	}
+
 	switch (accmode & O_ACCMODE) {
 	case O_RDONLY:
 		priv->push = 0;
-- 
2.7.4

